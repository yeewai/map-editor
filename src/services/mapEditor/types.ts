export interface State {
    zoom: number,
    panV: number,
    panH: number
}
