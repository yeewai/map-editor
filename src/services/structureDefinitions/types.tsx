export interface StructureDefinition {
    id: string,
    name: string,
    description?: string,
    width: number,
    length: number,
    kind: string,
    imageUrl: string,
    imageWidth?: number,
    imageHeight?: number,
    image?: any
}

export const emptyStructureDefintion = {
    id: "",
    name: "",
    width: 1,
    length: 1,
    kind: "",
    imageUrl: ""
}

export interface State {
    items: StructureDefinition[],
    itemsRequested: number,
    isFetching: boolean,
    hasFetched: boolean,
    error?: string
}
