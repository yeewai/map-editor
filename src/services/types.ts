import { State as structureDefinitionType } from './structureDefinitions/types';
import { State as worldType } from './worlds/types';

import { defaultState as structureDefinitionState } from './structureDefinitions/reducer';
import { defaultState as worldState } from './worlds/reducer';

export interface StateTree {
    structureDefinitions: structureDefinitionType,
    worlds: worldType
}

export const defaultStateTree: StateTree = {
    structureDefinitions: structureDefinitionState,
    worlds: worldState
}

export interface Action {
    type: string,
    payload?: any
}

export const defaultAction:Action = {
    type: ""
}
