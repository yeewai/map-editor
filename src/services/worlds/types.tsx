import { structureDefinitionTypes } from 'services/structureDefinitions';

export interface Structure {
    name: string,
    description?: string,
    definitionId: string,
    definition?: structureDefinitionTypes.StructureDefinition
    xposition: number,
    yposition: number
}

export const emptyStructure = {
    name: "",
    definitionId: structureDefinitionTypes.emptyStructureDefintion.id,
    definition: structureDefinitionTypes.emptyStructureDefintion,
    xposition: 1,
    yposition: 1
}

export interface World {
    id: string,
    name: string,
    key: string,
    description?: string,
    width: number,
    length: number,
    createdAt: number,
    saveMessage?: string,
    isPublished: boolean,
    structures?: Structure[],
    nullStructureId: string
}

export const emptyWorld = {
    id: "",
    name: "",
    key: "",
    width: 1,
    length: 1,
    createdAt: 1,
    isPublished: false,
    structures: [ emptyStructure ],
    nullStructureId: structureDefinitionTypes.emptyStructureDefintion.id
}

export interface WorldWithBoard extends World {
    board: Structure[][],
    uniqueStructures: structureDefinitionTypes.StructureDefinition[]
}

export const emptyWorldWithBoard = {
    ...emptyWorld,
    board: [
        [ emptyStructure ]
    ],
    uniqueStructures: [ structureDefinitionTypes.emptyStructureDefintion ]
}

export interface State {
    items: World[],
    isFetching: boolean,
    hasFetched: boolean,
    error?: string
}
