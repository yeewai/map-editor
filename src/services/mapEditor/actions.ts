import { ActionCreator } from 'redux';
import { Action } from 'services/types';

import { structureDefinitionTypes } from 'services/structureDefinitions';
import { worldTypes } from 'services/worlds';

// -------------------------------------------------------------
// "Camera" controls
// -------------------------------------------------------------
export const zoomIn: ActionCreator<Action> = () => ({
    type: 'mapEditor/ZOOMIN'
});

export const zoomOut: ActionCreator<Action> = () => ({
    type: 'mapEditor/ZOOMOUT'
});

export const zoomReset: ActionCreator<Action> = () => ({
    type: 'mapEditor/ZOOMRESET'
});

export const panLeft: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANLEFT'
});

export const panRight: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANRIGHT'
});

export const panUp: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANUP'
});

export const panDown: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANDOWN'
});

export const panReset: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANRESET'
});

// -------------------------------------------------------------
// Editing controls
// -------------------------------------------------------------
export const setStructureDefinition: ActionCreator<Action> = (sd: structureDefinitionTypes.StructureDefinition ) => ({
    type: 'mapEditor/SET_BUILD_STRUCTUREDEFINITION',
    payload: sd
});

// Gonna use this with keyboard controls probably
// export const unsetStructureDefinition: ActionCreator<Action> = () => ({
//     type: 'mapEditor/UNSET_BUILD_STRUCTUREDEFINITION'
// })

export const setMousePosition: ActionCreator<Action> = ( x: number, y: number ) => ({
    type: 'mapEditor/SET_MOUSE_POSITION',
    payload: { x, y }
});

export const mouseClick: ActionCreator<Action> = ( x: number, y: number ) => ({
    type: 'mapEditor/MOUSE_CLICK',
    payload: { x, y }
});


// -------------------------------------------------------------
// Map Editor Active World
// -------------------------------------------------------------
export const setActiveWorld: ActionCreator<Action> = ( world: worldTypes.World ) => ({
    type: 'mapEditor/SET_ACTIVE',
    payload: world
});
