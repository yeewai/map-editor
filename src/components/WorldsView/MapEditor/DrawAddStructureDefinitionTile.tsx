import * as React from 'react';
import { connect } from "react-redux";

import { StateTree } from 'services/types';
import { structureDefinitionTypes } from 'services/structureDefinitions';
import { mapEditorSelectors } from 'services/mapEditor';

import { DrawTile } from './DrawTile';

interface OwnProps {
    uniqueStructures: structureDefinitionTypes.StructureDefinition[]
}

interface StateProps {
    buildStructureDefinition?: structureDefinitionTypes.StructureDefinition,
    mousePosition?: { x: number, y: number }
}

export const mapStateToProps = (state: StateTree, ownProps: OwnProps): StateProps => ({
    buildStructureDefinition: mapEditorSelectors.getBuildStructureDefinition(state),
    mousePosition: mapEditorSelectors.getMousePosition(state)
});

export type PropTypes = OwnProps & StateProps;

export const DrawAddStructureDefinitionTile: React.SFC<PropTypes> = ( { uniqueStructures, buildStructureDefinition, mousePosition } ) => {
    if (!(buildStructureDefinition && mousePosition)) { return null; }
    const shouldDrawStructureDef = !!uniqueStructures.findIndex( sd => sd.id === buildStructureDefinition.id );

    return (
        <g className="AddStructure" pointerEvents="none">
            { shouldDrawStructureDef
                ? <defs className="buildingStructureReference">
                        <symbol id={buildStructureDefinition.id} key={buildStructureDefinition.id}
                            viewBox={`0 0 ${buildStructureDefinition.imageWidth} ${buildStructureDefinition.imageHeight}`}
                            dangerouslySetInnerHTML={{__html: buildStructureDefinition.image.innerHTML}}
                            preserveAspectRatio="xMinYMax"
                        />
                    </defs>
                : null
            }
            <DrawTile structureDefinition={buildStructureDefinition}
                    x={mousePosition.x} y={mousePosition.y} key={ `buildStructureDefinition`}/>

        </g>
    )
};

export default connect<StateProps, any, any>(mapStateToProps)(DrawAddStructureDefinitionTile);
