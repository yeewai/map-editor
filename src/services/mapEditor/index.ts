import reducer from "./reducer";
import * as mapEditorActions from "./actions";
import * as mapEditorSelectors from "./selectors";
import * as mapEditorTypes from "./types";

export { mapEditorActions, mapEditorSelectors, mapEditorTypes };
export default reducer;
