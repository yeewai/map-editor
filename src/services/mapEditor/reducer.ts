import { Action } from 'services/types';
import * as Types from './types';

export const defaultState: Types.State = {
    zoom: 1,
    panV: 0,
    panH: 0
}

export const ZOOMRATE = 1.5;
export const MAXZOOM = 10;
export const PANRATE = 1;

const mapEditorReducer = (state = defaultState, action: Action ): Types.State => {
    switch(action.type) {
        case 'mapEditor/ZOOMIN':
            return { ...state, zoom: Math.min(state.zoom * ZOOMRATE, MAXZOOM) };
        case 'mapEditor/ZOOMOUT':
            return { ...state, zoom: state.zoom / ZOOMRATE };
        case 'mapEditor/ZOOMRESET':
            return { ...state, zoom: defaultState.zoom };
        case 'mapEditor/PANUP':
            return {...state, panV: state.panV - PANRATE}
        case 'mapEditor/PANDOWN':
            return {...state, panV: state.panV + PANRATE}
        case 'mapEditor/PANLEFT':
            return {...state, panH: state.panH - PANRATE}
        case 'mapEditor/PANRIGHT':
            return {...state, panH: state.panH + PANRATE}
        case 'mapEditor/PANRESET':
            return {...state, panH: defaultState.panH, panV: defaultState.panV}
        default:
            return state;
    }
}


export default mapEditorReducer;
