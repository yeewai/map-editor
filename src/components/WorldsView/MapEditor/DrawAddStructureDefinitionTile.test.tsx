import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';

import { getMockRouterProps } from 'helpers/test/react-router';
import { DrawAddStructureDefinitionTile, mapStateToProps } from './DrawAddStructureDefinitionTile';
// import { worldTypes } from 'services/worlds';
import { defaultStateTree } from 'services/types';
import { mapEditorSelectors, mapEditorActions } from 'services/mapEditor';

Enzyme.configure({ adapter: new Adapter() });

describe("Map Editor", () => {
    let sandbox, wrapper, routerProps, world;

    beforeEach(() => {
        sandbox = sinon.createSandbox();
    });

    afterEach(() => {
        sandbox.restore();
    });

    it ("renders", () => {
        wrapper = shallow(<DrawAddStructureDefinitionTile uniqueStructures={["a"]} buildStructureDefinition={"a"} mousePosition={"b"} />);
        expect(wrapper).toMatchSnapshot();
    });

    it ("renders the symbol if it does not exist in uniqueStructures", () => {
        wrapper = shallow(<DrawAddStructureDefinitionTile uniqueStructures={[]} buildStructureDefinition={{image: {innerHTML: <circle />}}} mousePosition={"b"} />);
        expect(wrapper).toMatchSnapshot();
    });

    it ("does not render if there is no buildStructureDefinition or mousePosition", () => {
        wrapper = shallow(<DrawAddStructureDefinitionTile uniqueStructures={[]} buildStructureDefinition={null} mousePosition={"b"} />);
        expect(wrapper.get(0)).toBeNull();

        wrapper = shallow(<DrawAddStructureDefinitionTile uniqueStructures={[]} buildStructureDefinition={"a"} mousePosition={null} />);
        expect(wrapper.get(0)).toBeNull();
    })

    it ("maps state to props", () => {
        sandbox.stub(mapEditorSelectors, 'getBuildStructureDefinition').returns("a")
        sandbox.stub(mapEditorSelectors, 'getMousePosition').returns("b")

        expect(mapStateToProps(defaultStateTree)).toMatchObject({buildStructureDefinition: "a", mousePosition: "b"});
    })

});
