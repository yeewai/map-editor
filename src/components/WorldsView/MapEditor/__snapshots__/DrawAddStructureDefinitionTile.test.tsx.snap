// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map Editor renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    buildStructureDefinition="a"
    mousePosition="b"
    uniqueStructures={
        Array [
            "a",
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        null,
        <DrawTile
          structureDefinition="a"
          x={undefined}
          y={undefined}
/>,
      ],
      "className": "AddStructure",
      "pointerEvents": "none",
    },
    "ref": null,
    "rendered": Array [
      null,
      Object {
        "instance": null,
        "key": "buildStructureDefinition",
        "nodeType": "class",
        "props": Object {
          "structureDefinition": "a",
          "x": undefined,
          "y": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "g",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          <DrawTile
            structureDefinition="a"
            x={undefined}
            y={undefined}
/>,
        ],
        "className": "AddStructure",
        "pointerEvents": "none",
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": "buildStructureDefinition",
          "nodeType": "class",
          "props": Object {
            "structureDefinition": "a",
            "x": undefined,
            "y": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "g",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Map Editor renders the symbol if it does not exist in uniqueStructures 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    buildStructureDefinition={
        Object {
            "image": Object {
              "innerHTML": <circle />,
            },
          }
    }
    mousePosition="b"
    uniqueStructures={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <defs
          className="buildingStructureReference"
>
          <symbol
                    dangerouslySetInnerHTML={
                              Object {
                                        "__html": <circle />,
                                      }
                    }
                    id={undefined}
                    preserveAspectRatio="xMinYMax"
                    viewBox="0 0 undefined undefined"
          />
</defs>,
        <DrawTile
          structureDefinition={
                    Object {
                              "image": Object {
                                "innerHTML": <circle />,
                              },
                            }
          }
          x={undefined}
          y={undefined}
/>,
      ],
      "className": "AddStructure",
      "pointerEvents": "none",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <symbol
            dangerouslySetInnerHTML={
                        Object {
                                    "__html": <circle />,
                                  }
            }
            id={undefined}
            preserveAspectRatio="xMinYMax"
            viewBox="0 0 undefined undefined"
/>,
          "className": "buildingStructureReference",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "dangerouslySetInnerHTML": Object {
              "__html": <circle />,
            },
            "id": undefined,
            "preserveAspectRatio": "xMinYMax",
            "viewBox": "0 0 undefined undefined",
          },
          "ref": null,
          "rendered": null,
          "type": "symbol",
        },
        "type": "defs",
      },
      Object {
        "instance": null,
        "key": "buildStructureDefinition",
        "nodeType": "class",
        "props": Object {
          "structureDefinition": Object {
            "image": Object {
              "innerHTML": <circle />,
            },
          },
          "x": undefined,
          "y": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "g",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <defs
            className="buildingStructureReference"
>
            <symbol
                        dangerouslySetInnerHTML={
                                    Object {
                                                "__html": <circle />,
                                              }
                        }
                        id={undefined}
                        preserveAspectRatio="xMinYMax"
                        viewBox="0 0 undefined undefined"
            />
</defs>,
          <DrawTile
            structureDefinition={
                        Object {
                                    "image": Object {
                                      "innerHTML": <circle />,
                                    },
                                  }
            }
            x={undefined}
            y={undefined}
/>,
        ],
        "className": "AddStructure",
        "pointerEvents": "none",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <symbol
              dangerouslySetInnerHTML={
                            Object {
                                          "__html": <circle />,
                                        }
              }
              id={undefined}
              preserveAspectRatio="xMinYMax"
              viewBox="0 0 undefined undefined"
/>,
            "className": "buildingStructureReference",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "dangerouslySetInnerHTML": Object {
                "__html": <circle />,
              },
              "id": undefined,
              "preserveAspectRatio": "xMinYMax",
              "viewBox": "0 0 undefined undefined",
            },
            "ref": null,
            "rendered": null,
            "type": "symbol",
          },
          "type": "defs",
        },
        Object {
          "instance": null,
          "key": "buildStructureDefinition",
          "nodeType": "class",
          "props": Object {
            "structureDefinition": Object {
              "image": Object {
                "innerHTML": <circle />,
              },
            },
            "x": undefined,
            "y": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "g",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
